File: matlab/+smalltargetmotiondetectors/+core/+apgstmd_core/AttentionModule.m
classdef AttentionModule < smalltargetmotiondetectors.core.BaseCore
    % AttentionModule class for attention mechanism
    %   This class implements the attention mechanism module in the ApgSTMD.
    
    properties
        kernalSize = 17;
        zetaList = [2, 2.5, 3, 3.5];
        thetaList = [0, pi/4, pi/2, pi*3/4];
        alpha = 1;
    end
    
    properties(Hidden)
        attentionKernal;
    end
    
    methods
        function self = AttentionModule()
            % Constructor method
            % Initializes the AttentionModule object
            
            self = self@smalltargetmotiondetectors.core.BaseCore();
        end
    end
    
    methods
        function init(self)
            % Initialization method
            % Initializes the attention kernel
            
            import smalltargetmotiondetectors.tool.kernel.*;
            self.attentionKernal = create_attention_kernal(...
                self.kernalSize, ...
                self.zetaList, ...
                self.thetaList ...
                );
        end
        
        function attentionOpt = process(self, retinaOpt, predictionMap)
            % Processing method
            % Processes the retinaOpt and predictionMap to generate the
            % attention-optimal output
            
            [r, s] = size(self.attentionKernal);
            
            if isempty(predictionMap)
                attentionOpt = retinaOpt;
            else
                mapRetinaOpt = retinaOpt .* predictionMap;
                
                for i = 1:r
                    for j = 1:s
                        if j == 1
                            attentionResponseWithJ = ...
                                conv2(mapRetinaOpt, self.attentionKernal{i, 1}, 'same');
                        else
                            attentionResponseWithJ = min( ...
                                attentionResponseWithJ, ...
                                conv2(mapRetinaOpt, self.attentionKernal{i, j}, 'same') ...
                                );
                        end
                    end

                    if i == 1
                        attentionResponse = attentionResponseWithJ;
                    else
                        attentionResponse = ...
                            max(attentionResponse, attentionResponseWithJ);
                    end
                end
                
                attentionOpt = retinaOpt + ...
                    self.alpha * max(attentionResponse, 0);
            end

            self.Opt = attentionOpt;
        end
    end
end

File: python/smalltargetmotiondetectors/core/apgstmd_core/AttentionModule.py
import numpy as np
from smalltargetmotiondetectors.core.base_core import BaseCore
from smalltargetmotiondetectors.tool.kernel import create_attention_kernel

class AttentionModule(BaseCore):
    """
    AttentionModule class for attention mechanism.
    
    This class implements the attention mechanism module in the ApgSTMD.
    """
    
    def __init__(self):
        """
        Constructor method.
        
        Initializes the AttentionModule object.
        """
        super().__init__()
        self.kernal_size = 17
        self.zeta_list = [2, 2.5, 3, 3.5]
        self.theta_list = [0, np.pi/4, np.pi/2, np.pi*3/4]
        self.alpha = 1
        self.attention_kernel = None
    
    def init(self):
        """
        Initialization method.
        
        Initializes the attention kernel.
        """
        self.attention_kernel = create_attention_kernel(
            self.kernal_size,
            self.zeta_list,
            self.theta_list
        )
    
    def process(self, retina_opt, prediction_map):
        """
        Processing method.
        
        Processes the retina_opt and prediction_map to generate the
        attention-optimal output.
        """
        r, s = self.attention_kernel.shape
        
        if prediction_map is None:
            attention_opt = retina_opt
        else:
            map_retina_opt = retina_opt * prediction_map
            attention_response = None
            
            for i in range(r):
                for j in range(s):
                    if j == 0:
                        attention_response_with_j = np.convolve(
                            map_retina_opt,
                            self.attention_kernel[i, 0],
                            mode='same'
                        )
                    else:
                        attention_response_with_j = np.minimum(
                            attention_response_with_j,
                            np.convolve(
                                map_retina_opt,
                                self.attention_kernel[i, j],
                                mode='same'
                            )
                        )
                
                if i == 0:
                    attention_response = attention_response_with_j
                else:
                    attention_response = np.maximum(
                        attention_response,
                        attention_response_with_j
                    )
            
            attention_opt = retina_opt + self.alpha * attention_response
        
        self.Opt = attention_opt
